### ISSUE TITLE
Can't fetch categories with PostgreSQL

### ISSUE THREAD

**Issue by carltongibson** [sulu-io/sulu-standard#522]
Again with **PostgreSQL**. Getting an SQL error accessing this URL: 

    /admin.php/admin/api/categories?fields=id%2Cname&flat=true&locale=en&sortBy=depth&sortOrder=asc

Actual Error:

```
SQLSTATE[42803]: Grouping error: 7 ERROR:  column "c1_.translation" must appear in the GROUP BY clause or be used in an aggregate function
LINE 1: SELECT c0_.id AS id_0, c1_.translation AS translation_1, c0_...
```

This looks like the the approximate source:

```
at Sulu\Bundle\CategoryBundle\Controller\CategoryController>getCategoryListRepresentation(object(Request))
        in .../sulu-standard/vendor/sulu/sulu/src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php line 154
```

**Comment by wachterjohannes** [comment]
hmm it looks like that the listbuilder makes problems here ... maybe the line https://github.com/sulu-io/sulu/blob/f264d6213218caa421abd88e7f19924ee8b1a908/src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php#L307 because thats the main difference between tag and category with the listbuilder ...

**Comment by carltongibson** [comment]
Yes — that would make sense of the SQL error...  

> @carltongibson went off to play. 

... and, indeed, if I comment out that line the request loads and the categories are displayed. 

**But presumably** the `GROUP BY` is meant to be doing something — what's lost if we just delete it?

(And, it should work, I'd think)

**Comment by wachterjohannes** [comment]
for me the problem is not this line ... i have to test but i think we could perhaps remove it. the function group-by should be "usable" for postgres because if not we run always into the same problems with that (:

**Comment by carltongibson** [comment]
Exactly.

**Comment by carltongibson** [comment]
This [looks like a know issue with Doctrine + PostgreSQL](http://www.doctrine-project.org/jira/browse/DDC-2917). 

Dropping ordering and where clauses the SQL being generated looks like this: 

```
SELECT c0_.translation AS translation_0, c1_.id AS id_1, c1_.depth AS depth_2, c2_.id AS id_3, c3_.id AS id_4
FROM ca_categories c1_ 
LEFT JOIN ca_category_translations c0_ ON c1_.id = c0_.idCategories AND (c0_.locale = 'en') 
LEFT JOIN ca_categories c2_ ON c1_.idCategoriesParent = c2_.id 
LEFT JOIN ca_categories c3_ ON c1_.id = c3_.idCategoriesParent
GROUP BY c1_.id
```

What's missing is that all selected fields must appear in the GROUP BY clause. This works:

```
...
GROUP BY c1_.id, c0_.translation, c2_.id, c3_.id
```

Question: where to tell Doctrine to add the ids from the joined tables to the group by clause?

**Comment by wachterjohannes** [comment]
when i should be honest (: i have no idea! but i think this should be handled by doctrine ... maybe @danrot or @chirimoya have an idea!

**Comment by danrot** [comment]
I am not sure if I get this right, but why not just do it like this:

```php
$listBuilder->addSelectField($fieldDescriptors['depth']);
$listBuilder->addSelectField($fieldDescriptors['parent']);
$listBuilder->addSelectField($fieldDescriptors['hasChildren']);
$listBuilder->addGroupBy($fieldDescriptors['id']);
$listBuilder->addGroupBy($fieldDescriptors['parent']);
$listBuilder->addGroupBy($fieldDescriptors['hasChildren']);
```

**Comment by carltongibson** [comment]
@danrot — This doesn't work:

```
[Semantical Error] line 0, col 108 near 'SuluCategoryBundle:CategoryChildren.id': Error: 'SuluCategoryBundle:CategoryChildren' is not defined.
```

(Haven't had time to look further as yet.)

**Comment by danrot** [comment]
Sry, my mistake, the first `addGroupBy` should have `depth` as index. Is it possible that postgres demands all select fields to be in the group by statement? Would explain this issue to me.

**Comment by carltongibson** [comment]
> Is it possible that postgres demands all select fields to be in the group by statement? 

It needs at least the id for all the joined tables (which get added to the SELECT) to also be in the GROUP BY. In testing I was able to drop the `depth` but not the `translation` fields, as per the working example above.

**Comment by danrot** [comment]
I am afraid I can't really help you, since I don't know Postgres too well... Just propose a solution, and we'll have a look :smile:

**Comment by carltongibson** [comment]
No worries. :smile:. It's on my list. 

For the moment I've blocked out the `addGroupBy` call. As @wachterjohannes says, that's not a solution but the categories load and can be assigned. 

I'm working on the theme end for a couple of sites ATM. Once I reach a certain point there I can loop back and take a look at this more closely.

**Comment by carltongibson** [comment]
Right, I've finally got back to this... — It's the only Postgres specific failure in the `CategoryBundle`. 

**First approach** is to block out the `GROUP BY`. The tests then pass. So the question is, Do we need it? 

**Second approach** is take on the error. To wit, I have a look at `CategoryManager::getFieldDescriptors` and see, what best guess says is, the missing `translation` field under the key `name`. I add this to the query:

    $listBuilder->addSelectField($fieldDescriptors['name']);
    ...
    $listBuilder->addGroupBy($fieldDescriptors['name']);

but we still get an error

    "Doctrine\ORM\Query\QueryException": 
      "[Semantical Error] line 0, col 108 near 'SuluCategoryBundle:CategoryTranslation.translation':
       Error: 'SuluCategoryBundle:CategoryTranslation' is not defined."

I'll follow this up but is seems curious. It's defined here: `sulu/src/Sulu/Bundle/CategoryBundle/Entity/CategoryTranslation.php`. I need to look into what's going on exactly.

**Comment by danrot** [comment]
Well, if we really don't need the `GROUP BY` call, I would prefer to omit it, since it probably would also result in better performance...
